tests/test_config/test_config.py

```diff
@@ -14,11 +17,38 @@
     _Opt,
     check_confval_types,
     correct_copyright_year,
+    is_serializable,
 )
 from sphinx.deprecation import RemovedInSphinx90Warning
 from sphinx.errors import ConfigError, ExtensionError, VersionRequirementError
 
 
+def test_is_serializable():
+    # check that objects with circular references are correctly handled
+
+    a, b = [1], [2]
+    a.append(b)
+    b.append(a)
+    assert is_serializable(a)
+
+    x = {'a': 1, 'b': {'c': 1}}
+    x['b'] = x
+    assert is_serializable(x)
+
+    class _IgnoreExtend(frozenset):
+        def __or__(self, other):
```

Comment:
I did not bother extending the other possible methods (that could possibly not delegate to `__or__`) but since *I* know that I am using the recursive guard with `|`, I think it's fine. If `is_serializable` changes in the future (for whatever reason), the test would fail (like, someone decides to use `.union()` instead of `|` for the recursive guard).